### ISSUE TITLE
Counterintuitive indexing behaviour

### ISSUE THREAD

**Issue by gokceneraslan** [theislab/anndata#14]
Anndata `__getitem__` seems to exhibit a cross-product (similar to Python slices or `np.ix_()`) behaviour instead of numpy fancy indexing behaviour (so when it's indexed with 3 rows and 3 columns, it returns a 3x3 anndata, not 3 scalars as in fancy indexing).

 This is very useful and intuitive, in my opinion, because when users specify cell and gene indices they mean cell and gene filtering:

```python
adata = sc.datasets.paul15_raw()

print(adata[[0, 4, 10], ['Sfpi1']])
print(adata[[0, 4, 10], ['Sfpi1', 'Gata1', 'Fli1']])
```
![image](https://user-images.githubusercontent.com/1140359/37555310-c4c84e26-29e5-11e8-858f-fdf62ef509e2.png)

However, this is the case only if the col/row index dimensions are compatible in terms of broadcasting:

```python
print(adata[[0, 4, 10], ['Sfpi1', 'Gata1']])
```
![image](https://user-images.githubusercontent.com/1140359/37555368-81418be4-29e6-11e8-810e-cc0deb61efd7.png)

This is a bit difficult to understand :) Either all 3 cases should raise an exception or they should all perform cross-product-like slicing. What do you think?

**Comment by gokceneraslan** [comment]
Using cross-product indexing would also be consistent with dataframe loc/iloc indexing: 
`df.iloc[[0, 4, 10], [2, 3]]`

**Comment by falexwolf** [comment]
This now triggers an error... One might think about implementing this in the future...

**Comment by ivirshup** [comment]
Implemented in #164
